### ISSUE TITLE
Mirahc emits static variable access on an instance variable

### ISSUE THREAD

**Issue by felixvf** [mirah/mirah#380]
Consider this Mirah code

```mirah
class Foo
	
	def some_instance_method
		@foo = 1
	end
	
	def self.some_static_method
		puts foo
	end
end

Foo.some_static_method
```

This code should fail to compile, as the ```puts foo``` statement refers to a ```foo``` which should either be a (static) method or a local variable. There is neither a static or non-static method "foo" nor a local variable "foo". However, there is an instance variable "foo", which is completely inaccessible to the static method "some_static_method".

Not recognizing this impossibility, the compiler currently emits a "getstatic" bytecode instruction on instance variable "foo", which then fails during runtime as 

```
java.lang.IncompatibleClassChangeError: Expected static field Foo.foo
```

A fix for this bug should not try to detect such a situation too soon, as the Mirah AST is subject to transformations which may make illegal code later legal. However, the situation should be attempted to be detected, and if detected, and the compiler should emit a compilation error.

**Comment by baroquebobcat** [comment]
Yeah, I think this is due to the way the typer interacts with the type system. With the new fields used elements in scopes, I think we could make this better.
